Blackbox download crc error packet retry - version 2

etracer65 opened the pull request:
Related to https://github.com/betaflight/betaflight-configurator/issues/411

Revised version of the patches to enable retrying MSP blackbox download packets in the event of a CRC error.  Resolves the problem of the blackbox download freezing for some people when using jumbo frames in BF 3.1 or later.

Revised the patches to include a `crcError` flag in the callback response rather than using the dataCompressionType byte in the payload.  This will also enable other MSP message types to utilize the `crcError` flag to retransmit if appropriate (future implementation).

Tested on CP210x (including high baud rates) and VCP targets.

mikeller commented on js/msp/MSPHelper.js:
```diff
@@ -957,7 +958,7 @@ MspHelper.prototype.process_data = function(dataHandler) {
 
             // fire callback
-            if (callback) callback({'command': code, 'data': data, 'length': data.byteLength});
+            if (callback) callback({'command': code, 'data': data, 'length': data.byteLength, 'crcError': crcError});
```
Who is consuming the crcError in this callback?

etracer65 replied:
The callback from `MspHelper.prototype.dataflashRead` in js/msp/MSPHelper.  But now that it's in the response object all the MSP message callbacks could use it for future implementations.  There might be other message types that could also benefit from knowing about a CRC error and retry.

But now that I'm thinking about it, it might make sense to put some logic in js/msp.js to only set the flag for specific message types that expect to deal with crc errors.  Because before the message payload was dropped (nulled) if a crc error occurred.  The callback was still called and the handler couldn't tell that there was actually a crc error and this caused the javascript runtime errors since the dataflash callback wasn't expecting the payload object to be null.  The difference is now that the payload is still returned even though it fails the crc check.  This might cause problems to other msp message type handlers in the (unlikely) event they get a crc error.

Maybe I should revise this slightly to still null out the payload (like the old crc handling), keep the flag to pass to the callbacks, and then refactor the `MspHelper.prototype.dataflashRead` callback to check the `crcError` flag before accessing the payload.  Thoughts?

etracer65 replied:
But frankly all of the callbacks and handlers for the msp messages are simply expecting the data object to be valid.  So any crc error regardless of the message type would case a runtime error.  It's just that crc errors are apparently rare with the exception of the new jumbo-packets.  This probably explains the very occasional Configurator hangs I've seen in the past.  Probably a crc error not being handled for some message resulting in a javascript runtime error.  Ideally every message handler *should* check for the new crcError flag and act appropriately but not sure if it's worth the effort.

etracer65 replied:
So to simplify…

Previous behavior: CRC error would result in the dataView being a null object and the callbacks would generate runtime errors because they tried to reference the object and had no idea that a crc error occurred.  This is the cause of the hang during blackbox downloads, but it would affect any of the msp message types.

Current patch:  Provides a crcError flag so the handlers can react appropriately and returns the received payload even though it has crc errors.  This works around the problem of the handlers referencing the null object, but this may not be a good idea as they may be acting on invalid data.  Previously they would crash, now they won't but would still have no idea that the data was invalid (except for blackbox downloading since it examines the crcError now).

Proposal: Revert the behavior of nulling the data view object and refactor the callback in `MspHelper.prototype.dataflashRead` to check the crcError flag before referencing the possibly null data object.  This would still allow any other message type to potentially crash on a crc error (the previous behavior) but eliminates the risk of them acting on potentially corrupt data.  Later each hander could be updated to examine the ecrError flag and act appropriately (drop the message or retry).

mikeller commented:
I reverted this for now, since it has the potential to get bogus data used in the configurator. I am not so much worried about bogus data in dataflash downloads (worst case the log is corrupted), but about bogus data used and then written back as part of the config, which could really ruin somebody's day.

My suggestion for a fix would be to extend the callback list to include a boolean callbackOnError, with dataflash registering its callback with this true, and then only calling back on handlers that were registered with this true for CRC errors.

etracer65 commented:
@mikeller Yes, I agree about the chance of corrupted data.  Note however that at the moment if any crc error occurs for even other msp messages that we'll get a javascript runtime error.  This is because the crc detection in js/msp.js is setting the DataView object to null and `MspHelper.prototype.process_data` will blindly use the object - causing the runtime errors.

I was in the middle of typing another reply when you sent this about some revisions I've made - let me post that separately for review.  It addresses most of the issues with the possibility of bogus data.  I like the idea of including a boolean callbackOnError so I'll look into this as well.

etracer65 commented:
@mikeller Before I generate another pull request, take a look at https://github.com/etracer65/betaflight-configurator/commit/6af472a4879e01e0b8ae94a637aac5bf0dd26219

I started looking deeper at the MSP message processing and see that CRC errors are a big problem for every message type.  The issue comes from js/msp.js setting the DataView to null on a CRC error.  Then `MspHelper.prototype.process_data` tries to process the payload and blindly references the DataView not knowing that it's invalid and we get a runtime error.  This is what was happening for the blackbox messages, but it would happen to *any* message in the event of a crc error.  As I said earlier, I'm sure this is the cause of intermittent Configurator freezes during normal use that I've seen on the same platforms that have the blackbox download problems - they're just getting periodic crc errors for other message types as well.

So the workaround in the first patch was to not invalidate the DataView object and return the received payload even though it is corrupted.  Then unfortunately the handlers in `MspHelper.prototype.process_data` would just process the data and that could lead to unexpected results or corrupted settings.  Setting the DataView to an empty object would be just as bad as the handlers once again would just blindly process the payload and load everything up with zeros.

So the *best* solution would be to put code in *every* message type handler in `MspHelper.prototype.process_data` to first check for a crcError condition before accessing the DataView and respond appropriately.  But this is not practical as you'd have to make strategic decisions about each message type to decide what to do.

So basically I wrapped a check around the message handlers in `MspHelper.prototype.process_data` to skip the processing if crcError is true.  If a callback was supplied on the MSP.send_message then the callback will still happen.  We need this to happen as this is where the retrying of the blackbox packet occurs.

I'm not sure what github did with the diff in js/msp/MSPHelper.js.  I wrapped a
`    if (!crcError) {`
around the case statements to prevent them processing the bogus data.

I tested this by putting in some test code in js/msp.js to randomly trigger false crc errors at up to a 10% rate (random message types) and the configurator continued to perform normally.

I'll look into your other suggestion a boolean to prevent the callback on crc errors unless the requester sets the flag.  I think adding that along with the changes so far will make everything solid.

mikeller commented:
Yes, making `MspHelper.prototype.process_data` aware of a CRC error, and make it not do any updates is a good approach. If we can make reasonably sure that we put the same error awareness in all other callbacks then my suggestion to prevent callbacks for non error aware callbacks (I don't think there are many more, but we need to have a close look) is not needed really.

One more thing: We should rename the `crcError` that is passed back into `error`, since it could / should be used to signal other error conditions as well.

etracer65 commented:
@mikeller Ok, I revised the patch to include logic to prevent the callback on crc errors unless the requester specifically requires it.  Currently the only message using it is blackbox downloads.  The flag is added as an optional parameter in MSP.send_message and must be set to true to request the callback on crc errors.  The parameter defaults to false if excluded so that all the existing calls will not request the callback on crc errors.  This should completely eliminate any vulnerabilities to crc errors for all message types.  The worst case scenario is a timeout if the requested response never comes from the flight controller because the crc error was dropped.  Of course the dataflash download is modified to catch this and retry.

Review the changes here before I submit a pull request to see if you think I'm on the right track.
https://github.com/etracer65/betaflight-configurator/commit/53c0bb8ca9fb6e270d7c87d087ce2c048f525709

I confirmed that blackbox downloads work reliably for both CP210x and VCP targets and successfully recover from crc errors.  Additionally I stress tested by inserting random crc errors (with debug code) and confirmed that the configurator behaved normally even with 100's of crc errors.

etracer65 commented:
@mikeller Sorry, didn't see your reply.  I'll change `crcError` to `error` in the callback response.  I already made the changes to support the `callbackOnError` parameter when the callback is registered and not execute the callback when the error occurs.  So this should cover everything.  Let me rename the variables and I'll give you another link to review.

etracer65 commented:
@mikeller Actually, on the change from `crcError`…  I understand what you're saying about potentially returning other errors, but I think it's important to differentiate the crc error.  For example the blackbox download will want to retry crc errors but possibly not other errors.  Currently there are no other errors that could be returned, but I guess you're planning for the future.

etracer65 commented:
Had to re-sync my repository.  Working version before the possible crcError name change is here for review:

https://github.com/etracer65/betaflight-configurator/commit/8a286bdeed4358a0b55d3fffcffecbf4ab5459c4

mikeller commented:
Looks good to me, a few minor comments.

etracer65 commented:
Ok, thanks.  I've made the small corrections and ran though my test suite again and everything was good.  I'll submit a pull request shortly.
